### PR TITLE
calling flows from flows and other support for 9796

### PR DESCRIPTION
**Pull request by turadg**
refs: #9796 

## Description

https://github.com/Agoric/agoric-sdk/pull/10023/ started with a series of refactors and cleanup. This separates those into another PR to be reviewed more rapidly.

It also includes a new feature to orchestrateAll to ease calling flows from flows. Each flow is available on the context object under `flows`.

### Security Considerations
Every flow can call another flow. In the event a developer wants isolate between flows, they won't put them in the same `orchestrateAll`.

### Scaling Considerations
n/a

### Documentation Considerations
none

### Testing Considerations
new tests

### Upgrade Considerations
not yet deployed

### CODE REVIEW

File: packages/async-flow/src/endowments.js [hunk 1]
```diff
@@ -64,7 +64,7 @@ export const forwardingMethods = rem => {
  * @param {R} dataRecord
  * @returns {R}
  */
-export const makeStateRecord = dataRecord =>
+export const makeSharedStateRecord = dataRecord =>
```
Comment (0xpatrickdev) [inline packages/async-flow/src/endowments.js:67]: Nice - I erroneously thought this had already landed

File: packages/orchestration/src/examples/unbond.flows.js [hunk 2]
```diff
@@ -24,13 +24,12 @@ export const unbondAndLiquidStake = async (
   _offerArgs,
 ) => {
   console.log('zcf within the membrane', zcf);
-  // We would actually alreaady have the account from the orchestrator
-  // ??? could these be passed in? It would reduce the size of this handler,
-  // keeping it focused on long-running operations.
-  const omni = await orch.getChain('omniflixhub');
-  const omniAccount = await omni.makeAccount();
+  // Osmosis is one of the few chains with icqEnabled
+  const osmosis = await orch.getChain('osmosis');
```
Comment (0xpatrickdev) [inline packages/orchestration/src/examples/unbond.flows.js:28]: `omniflix` appears to have icq enabled, but that's not to suggest we should revert this change. `osmosis` is preferable from my POV since it's integrated in `multichain-testing` - starship doesn't yet support `omniflix`.

File: packages/orchestration/src/exos/portfolio-holder-kit.js [hunk 3]
```diff
@@ -43,7 +43,7 @@ const preparePortfolioHolderKit = (zone, { asVow, when }) => {
     'PortfolioHolderKit',
     {
       invitationMakers: M.interface('InvitationMakers', {
-        MakeInvitation: M.call(
+        Proxying: M.call(
```
Comment (0xpatrickdev) [inline packages/orchestration/src/exos/portfolio-holder-kit.js:46]: Wb `Proxy`, verb, instead of a present participle?

File: packages/orchestration/test/facade-durability.test.ts [hunk 4]
```diff
@@ -1,15 +1,15 @@
 import { test as anyTest } from '@agoric/zoe/tools/prepare-test-env-ava.js';
 
+import { makeIssuerKit } from '@agoric/ertp';
+import { reincarnate } from '@agoric/swingset-liveslots/tools/setup-vat-data.js';
```
Comment (0xpatrickdev) [inline packages/orchestration/test/facade-durability.test.ts:4]: Have we considered adopting eslint `sort-imports`? Feel like I've seen this type of change frequently in recent reviews

**Review by 0xpatrickdev — APPROVED** [review]
LGTM - please consider the `Proxying` -> `Proxy` name change

File: packages/orchestration/src/exos/portfolio-holder-kit.js (hunk 3 above)
Comment (turadg) [reply packages/orchestration/src/exos/portfolio-holder-kit.js:46]: That was my first thought too but since [Proxy is a global object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) I thought it would be confusing.

File: packages/orchestration/test/facade-durability.test.ts (hunk 4 above)
Comment (turadg) [reply packages/orchestration/test/facade-durability.test.ts:4]: We have :) https://github.com/Agoric/agoric-sdk/issues/7403

Feel free to lobby to prioritize it.

File: packages/orchestration/src/facade.js [hunk 5]
```diff
@@ -106,16 +108,46 @@ export const makeOrchestrationFacade = ({
...
+          allGuestFns.has(value) ? mappedGuestFns[key] : value,
+        ]),
+      );
+
+    // TODO support arbitrary location of guest fns in context
+    // https://github.com/Agoric/agoric-sdk/issues/9823
+    const mappedContext = Object.fromEntries(
+      Object.entries(hostCtx).map(([key, value]) => [
+        key,
+        Object.entries(value).length > 0 ? convertGuests(value) : value,
```
Comment (mhofman) [inline packages/orchestration/src/facade.js]: This will break if `ctx` is a of the shape `{ farOrExo }`,

**Review by mhofman — DISMISSED** [review]

File: packages/internal/src/utils.js [hunk 6]
```diff
@@ -52,6 +52,48 @@ export const deeplyFulfilledObject = async obj => {
...
+    );
+    wasMapped ||= mappedInnerValue !== innerValue;
+    return [innerName, mappedInnerValue];
+  });
+
+  return wasMapped ? Object.fromEntries(mappedEntries) : value;
+};
+
+/**
+ * @param {object} obj
```
Comment (turadg) [inline packages/internal/src/utils.js:97]: please include some documentation.

a parameterized typedef would be nice but not worth the effort yet imo

File: packages/internal/test/utils.test.js [hunk 7]
```diff
@@ -30,6 +31,99 @@ test('deeplyFulfilledObject', async t => {
...
+        inputVal,
+        `${['obj', ...unchangedPath].join('.')} is unchanged`,
+      );
+    }
+
+    if (expectedOutput) {
+      t.deepEqual(output, expectedOutput);
+    }
+  },
+  title(providedTitle, input) {
```
Comment (turadg) [inline packages/internal/test/utils.test.js]: nice macro! consider putting title up top

File: packages/orchestration/src/facade.js [hunk 8]
```diff
@@ -109,25 +109,16 @@ export const makeOrchestrationFacade = ({
...
-        // https://github.com/Agoric/agoric-sdk/issues/9823
-        value === guestFns ? getMappedFlows() : value,
+    const mappedFlows = new Map(
+      Object.entries(guestFns).map(([name, guestFn]) => [
+        guestFn,
+        // eslint-disable-next-line no-use-before-define
+        (...args) => orcFns[name](...args),
       ]),
     );
+    const mapFlow = guestFn => mappedFlows.get(guestFn) || guestFn;
```
Comment (turadg) [inline packages/orchestration/src/facade.js]: a little awkward that `mapFlow` uses `mappedFlows`.

consider inlining this fn

File: packages/internal/src/utils.js (hunk 6 above)
Comment (mhofman) [reply packages/internal/src/utils.js:97]: I couldn't think how a parametrized typedef would work in this case. At best we could pretend that the output type is the same as the input type.

Comment (turadg) [reply packages/internal/src/utils.js:97]: I was thinking it would maintain the structure but the return type of each leaf would be widened to the return type of the mapping function. (Since the type system doesn't know more about what the function will do)

But still not worth it

**Review by mhofman — COMMENTED** [review]
I added some commits to enable mapping guest flows anywhere in the context object and updated a test accordingly.
